README.md
# Easily convert images with Glide Image Utility.

[![Latest Version on Packagist](https://img.shields.io/packagist/v/mindinventory/glide-image-utility.svg?style=flat-square)](https://packagist.org/packages/mindinventory/glide-image-utility)
[![GitHub Tests Action Status](https://img.shields.io/github/actions/workflow/status/mindinventory/glide-image-utility/run-tests.yml?branch=main&label=tests&style=flat-square)](https://github.com/mindinventory/glide-image-utility/actions?query=workflow%3Arun-tests+branch%3Amain)
[![GitHub Code Style Action Status](https://img.shields.io/github/actions/workflow/status/mindinventory/glide-image-utility/fix-php-code-style-issues.yml?branch=main&label=code%20style&style=flat-square)](https://github.com/mindinventory/glide-image-utility/actions?query=workflow%3A"Fix+PHP+code+style+issues"+branch%3Amain)
[![Total Downloads](https://img.shields.io/packagist/dt/mindinventory/glide-image-utility.svg?style=flat-square)](https://packagist.org/packages/mindinventory/glide-image-utility)

This package provides an easy-to-use class to manipulate images. Using this package we can manipulate aws s3 bucket image also.

## Installation

You can install the package via composer:

```bash
composer require mindinventory/glide-image-utility
```

You can publish the config file with:

```bash
php artisan vendor:publish --tag="glide-image-utility-config"
```

This is the contents of the published config file:

```php
return [

];
```

## Usage

```php
MiImage::createImage($ImagePath,['w'=> 50, 'h'=>100, 'fit'=>'crop', 'bg' => 'CCC']);
```

### In modification parameter you can use

**w** = Width,

**h** = Height,

**fit** = Fit parameter [contain, max, fill, stretch, crop],

**q** = Quality [Between 0 to 100],

**fm** = Format [jpg, pjeg(progressive jpg), gif, webp]

**bg** = Background Color [For More Colors](https://glide.thephpleague.com/1.0/api/colors/)

## Testing

```bash
composer test
```

## Changelog

Please see [CHANGELOG](CHANGELOG.md) for more information on what has changed recently.

## Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) for details.

## Security Vulnerabilities

Please review [our security policy](../../security/policy) on how to report security vulnerabilities.

## Credits

- [MindInventory](https://github.com/Mindinventory)
- [All Contributors](../../contributors)

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.

## Let us know!
If you use our open-source libraries in your project, please make sure to credit us and Give a star to [www.mindinventory.com](https://mindinventory.com/)


![alt text](https://git.mindinventory.com/uploads/-/system/appearance/header_logo/1/mi-logo.png)

<a href="https://www.mindinventory.com/contact-us.php?utm_source=gthb&utm_medium=repo&utm_campaign=glide-image-utility" target="__blank">
<img src="https://github.com/Sammindinventory/MindInventory/raw/main/hirebutton.png" width="203" height="43"  alt="app development">
</a>


CHANGELOG.md
# Changelog

All notable changes to `glide-image-utility` will be documented in this file.

> v2.0.0
- Added : PHP 8.1 support
- Added : Laravel 10 support
- Upgrade : [spatie/laravel-glide](https://packagist.org/packages/spatie/laravel-glide) version to `3.7.*`
- Added : Test case

> v1.0.5
- Added : PHP 8.0 support
- Added : Laravel 8 support
- Upgrade : [spatie/laravel-glide](https://packagist.org/packages/spatie/laravel-glide) version to `3.6.*`

> v1.0.4
- Upgrade : [spatie/laravel-glide](https://packagist.org/packages/spatie/laravel-glide) version to `3.5.*`
- Updated : [README](./README.md)

> v1.0.3
- Bug fixed
- Added : Laravel 7 support
- Updated : [README](./README.md)

> v1.0.2
- Bug fixed

> v1.0.1
- Added : Background color support
- Updated : [README](./README.md)

> v1.0.0
- Convert images with Glide Image Utility


tests/Pest.php
<?php

use Mi\MiImageUtility\Tests\TestCase;

uses(TestCase::class)->in(__DIR__);


tests/TestCase.php
<?php

namespace Mi\MiImageUtility\Tests;

use Illuminate\Database\Eloquent\Factories\Factory;
use Orchestra\Testbench\TestCase as Orchestra;
use Mi\MiImageUtility\MiImageUtilityServiceProvider;

class TestCase extends Orchestra
{
    protected function setUp(): void
    {
        parent::setUp();
    }

    protected function getPackageProviders($app)
    {
        return [
            MiImageUtilityServiceProvider::class,
        ];
    }

    public function getEnvironmentSetUp($app)
    {
        config()->set('database.default', 'testing');
    }
}


tests/BaseCasesTest.php
<?php

use Illuminate\Support\Facades\Storage;
use Mi\MiImageUtility\MiImageUtility;
use Symfony\Component\HttpFoundation\BinaryFileResponse;

beforeEach(function () {
    $this->image = fake()->imageUrl();
});

it('invalid image', function () {
    $imageData = MiImageUtility::createImage(
        Storage::disk('public')->path('mi-image-test-4.png'),
        ['w' => 100, 'h' => 50, 'fit' => 'crop', 'bg' => 'CCC']
    );
    expect($imageData)->not()->toBeInstanceOf(BinaryFileResponse::class);
});

it('image from public path', function () {
    copy($this->image, public_path('mi-image-test.png'));
    $imageData = MiImageUtility::createImage(
        public_path('mi-image-test.png'),
        ['w' => 100, 'h' => 50, 'fit' => 'crop', 'bg' => 'CCC']
    );
    expect($imageData)->toBeInstanceOf(BinaryFileResponse::class);
    unlink(public_path('mi-image-test.png'));
});

it('reduce image quality to 50', function () {
    copy($this->image, public_path('mi-image-test-1.png'));
    $imageData = MiImageUtility::createImage(
        public_path('mi-image-test-1.png'),
        ['w' => 100, 'h' => 50, 'fit' => 'crop', 'q' => 50, 'bg' => 'CCC']
    );
    expect($imageData)->toBeInstanceOf(BinaryFileResponse::class);
    unlink(public_path('mi-image-test-1.png'));
});

it('change image format from png to jpg', function () {
    copy($this->image, public_path('mi-image-test-2.png'));
    $imageData = MiImageUtility::createImage(
        public_path('mi-image-test-2.png'),
        ['w' => 100, 'h' => 50, 'fit' => 'crop', 'q' => 50, 'fm' => 'jpg', 'bg' => 'CCC']
    );
    expect($imageData)->toBeInstanceOf(BinaryFileResponse::class);
    unlink(public_path('mi-image-test-2.png'));
});

it('image from storage path', function () {
    copy($this->image, Storage::disk('public')->path('mi-image-test-3.png'));
    $imageData = MiImageUtility::createImage(
        Storage::disk('public')->path('mi-image-test-3.png'),
        ['w' => 100, 'h' => 50, 'fit' => 'crop', 'bg' => 'CCC']
    );
    expect($imageData)->toBeInstanceOf(BinaryFileResponse::class);
    Storage::disk('public')->delete('mi-image-test-3.png');
});

it('invalid image type', function () {
    $imageData = MiImageUtility::createImage(
        'https://www.mindinventory.com/static/media/logo-mind-inventory-w.8514e2e8.svg',
        ['w' => 100, 'h' => 50, 'fit' => 'crop', 'fm' => 'png', 'bg' => 'CCC']
    );
    expect($imageData)->not()->toBeInstanceOf(BinaryFileResponse::class);
});

it('remote image', function () {
    $imageData = MiImageUtility::createImage(
        'https://www.google.com/images/branding/googlelogo/1x/googlelogo_color_272x92dp.png',
        ['w' => 100, 'h' => 50, 'fit' => 'crop', 'fm' => 'jpg', 'bg' => 'CCC']
    );
    expect($imageData)->toBeInstanceOf(BinaryFileResponse::class);
});




src/MiImageUtility.php
<?php

namespace Mi\MiImageUtility;

use Exception;
use Illuminate\Http\JsonResponse;
use Illuminate\Http\Response;
use Spatie\Glide\GlideImage;
use Symfony\Component\HttpFoundation\BinaryFileResponse;

class MiImageUtility
{
    /***
     * @param $file
     * @param array $modificationParameters
     * @return JsonResponse|BinaryFileResponse
     */
    public static function createImage($file, array $modificationParameters): BinaryFileResponse|JsonResponse
    {

        $info = pathinfo($file);
        $img = $info['basename'];
        $i = explode('.', $img);

        if (!$img || !is_string($file)) {
            return response()->json(["data" => "Invalid Image"], Response::HTTP_NOT_FOUND);
        } else if ($i[count($i) - 1] == 'jpg' || $i[count($i) - 1] == 'png' || $i[count($i) - 1] == 'jpeg') {

            $width = '';
            if (isset($modificationParameters['w'])) {
                $width = $modificationParameters['w'];
            }
            $height = '';
            if (isset($modificationParameters['h'])) {
                $height = $modificationParameters['h'];
            }
            $fit = '';
            if (isset($modificationParameters['fit']) && in_array($modificationParameters['fit'], ['contain', 'max', 'fill', 'stretch', 'crop'])) {
                $fit = $modificationParameters['fit'];
            }
            $quality = '';
            if (isset($modificationParameters['q']) && $modificationParameters['q'] <= 100 && $modificationParameters['q'] > 0) {
                $quality = (int)$modificationParameters['q'];
            }
            $format = '';
            if (isset($modificationParameters['fm']) && in_array($modificationParameters['fm'], ['jpg', 'pjpg', 'png', 'gif', 'webp'])) {
                $format = $modificationParameters['fm'];
            }
            $bg = '';
            if (isset($modificationParameters['bg'])) {
                $bg = $modificationParameters['bg'];
            }

            $tmpImageArr = (explode('.', $img));
            $ext = end($tmpImageArr);
            $image_name = basename($img, '.' . $ext);
            $cacheImage = $image_name . '-' . (($width) ? $width : 0) . 'x' . (($height) ? $height : 0) . $fit . $quality . $bg;
            $cacheImage = $cacheImage . '.' . $ext;

            $newImage = self::checkFolderAvailability($file,$cacheImage);

            if (file_exists(storage_path($newImage))) {
                return new BinaryFileResponse(storage_path($newImage), 200);
            } else {
                $files = storage_path('images/' . $info['basename']);

                try {
                    $contents = file_get_contents($file);
                    file_put_contents($files, $contents);
                }catch (Exception $e){
                    return response()->json(["data" => "Invalid Image"], Response::HTTP_NOT_FOUND);
                }

                $pathToImage = storage_path('images/' . $img);

                GlideImage::create($pathToImage)
                    ->modify(['w' => $width, 'h' => $height, 'fit' => $fit, 'q' => $quality, 'fm' => $format, 'bg' => $bg])
                    ->save(storage_path($newImage));
                chmod(storage_path($newImage), 0777);
                unlink($pathToImage);

                return new BinaryFileResponse(storage_path($newImage), 200);
            }
        } else {
            return response()->json(["data" => "Invalid Image"], Response::HTTP_NOT_FOUND);
        }
    }

    /***
     * @param $file
     * @param $cacheImage
     * @return string
     */
    private static function checkFolderAvailability($file, $cacheImage): string
    {
        $path = parse_url($file, PHP_URL_PATH);
        $items = array_filter(explode('/', $path),fn($val) => $val);
        $folder = "";
        $newImage = 'images/cache/' . $cacheImage;
        if (count($items) > 2) {
            $folder = $items[count($items) - 2];
            $newImage = 'images/cache/' . $folder . '/' . $cacheImage;
        }

        $imageFolder = 'images/cache/' . $folder;
        if (!is_dir(storage_path($imageFolder))) {
            mkdir(storage_path($imageFolder), 0777, true);
            chmod(storage_path($imageFolder), 0777);
        }
        return $newImage;
    }
}


src/MiImage.php
<?php

namespace Mi\MiImageUtility;

use Illuminate\Http\Response;
use Spatie\Glide\GlideImage;
use Symfony\Component\HttpFoundation\BinaryFileResponse;

class MiImage
{
    public static function createImage($file, array $modificationParameters)
    {
        $info = pathinfo($file);
        $img = $info['basename'];
        $i = explode('.', $img);

        if (!$img) {
            return response()->json(["data" => "Invalid Image"], Response::HTTP_NOT_FOUND);
        } else if ($i[count($i) - 1] == 'jpg' || $i[count($i) - 1] == 'png' || $i[count($i) - 1] == 'jpeg') {

            $width = '';
            if (isset($modificationParameters['w'])) {
                $width = $modificationParameters['w'];
            }
            $height = '';
            if (isset($modificationParameters['h'])) {
                $height = $modificationParameters['h'];
            }
            $fit = '';
            if (isset($modificationParameters['fit']) && in_array($modificationParameters['fit'], ['contain', 'max', 'fill', 'stretch', 'crop'])) {
                $fit = $modificationParameters['fit'];
            }
            $quality = '';
            if (isset($modificationParameters['q']) && $modificationParameters['q'] <= 100 && $modificationParameters['q'] > 0) {
                $quality = (int)$modificationParameters['q'];
            }
            $format = '';
            if (isset($modificationParameters['fm']) && in_array($modificationParameters['fm'], ['jpg', 'pjpg', 'png', 'gif', 'webp'])) {
                $format = $modificationParameters['fm'];
            }
            $bg = '';
            if (isset($modificationParameters['bg'])) {
                $bg = $modificationParameters['bg'];
            }

            $tmpImageArr = (explode('.', $img));
            $ext = end($tmpImageArr);
            $image_name = basename($img, '.' . $ext);
            $cacheImage = $image_name . '-' . (($width) ? $width : 0) . 'x' . (($height) ? $height : 0) . $fit . $quality . $bg;
            $cacheImage = $cacheImage . '.' . $ext;

            $items = explode('/', $file);
            if (sizeof($items) > 1) {
                $folder = $items[sizeof($items) - 2];
                $newImage = 'images/cache/' . $folder . '/' . $cacheImage;
            } else {
                $folder = "";
                $newImage = 'images/cache/' . $cacheImage;
            }


            if (file_exists(storage_path($newImage))) {
                return new BinaryFileResponse(storage_path($newImage), 200);
            } else {

                $contents = file_get_contents(storage_path('images/'.$file));
                $image = storage_path('images');
                if (!is_dir($image)) {
                    mkdir($image, 0777, true);
                    chmod($image, 0777);
                }

                $file = storage_path('images/' . $info['basename']);
                file_put_contents($file, $contents);
                chmod($file, 0777);

                $imageFolder = storage_path('images/cache/' . $folder);
                if (!is_dir($imageFolder)) {
                    mkdir($imageFolder, 0777, true);
                    chmod($imageFolder, 0777);
                }

                $pathToImage = storage_path('images/' . $img);

                GlideImage::create($pathToImage)
                    ->modify(['w' => $width, 'h' => $height, 'fit' => $fit, 'q' => $quality, 'fm' => $format, 'bg' => $bg])
                    ->save(storage_path($newImage));
                chmod(storage_path($newImage), 0777);
                unlink($pathToImage);

                return new BinaryFileResponse(storage_path($newImage), 200);
            }
        } else {
            return response()->json(["data" => "Invalid Image"], Response::HTTP_NOT_FOUND);
        }
    }
}
